File: week-4/partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums) %2 != 0: return False
        hashtable = {}
        return self.backtrack(0, 0, sum(nums), nums, hashtable)
    
    def backtrack(self, index, cur_sum, total, nums, hashtable):
        state = str(cur_sum) + '~~' + str(index)
        if state in hashtable:
            return hashtable[state]
        
        elif cur_sum*2 == total:
            hashtable[state] = True
        
        elif cur_sum > total/2 or index >= len(nums):
            hashtable[state] = False
    
        else:
            hashtable[state] = self.backtrack(index+1, cur_sum, total, nums, hashtable) or self.backtrack(index+1, cur_sum+nums[index], total, nums, hashtable)
            
        return hashtable[state]

File: week-4/partition-equal-subset-sum/partition-equal-subset-sum.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var canPartition = function(nums) {
    const sum = nums.reduce((a,b)=>a+b, 0);
    if (sum %2 !== 0) return false;
    const hashtable = {};
    return backtrack(0, 0, sum, nums, hashtable);
};

const backtrack = (index, cur_sum, total, nums, hashtable) => {
    const state = cur_sum + '~~' + index;
    if (state in hashtable) return hashtable[state];
    
    if (cur_sum*2 === total) hashtable[state] = true;
    
    else if (cur_sum > total || index >= nums.length) hashtable[state] = false;
    
    else
        hashtable[state] = backtrack(index+1, cur_sum, total, nums, hashtable) || backtrack(index+1, cur_sum+nums[index], total, nums, hashtable);
    
    return hashtable[state];
}
